### ISSUE TITLE
[REQUEST] Allow connecting to a remote host with a unix socket

### ISSUE THREAD

**Issue by baikunz** [amir20/dozzle#2465]
**Is your feature request related to a problem? Please describe.**
I need to connect to a remote docker host without exposing its API (which is not exactly straightforward to setup securely)
I would like to be able to use a unix socket tunnelled through SSH to connect to the remote host.

**Describe the solution you'd like**
We are already connecting to the local docker unix socket for the local host.
I'd like to be able to connect to a remote host with another socket instead of only tcp.

**Describe alternatives you've considered**
I looked at exposing the API but securing it properly is way more cumbersome that I would like compared to the simplicity of an SSH tunnel.

**Comment by amir20** [comment]
@baikunz to be honest, when I first implemented remote connections, I started with SSH. However, through a lot of testing, I found out that ssh is very slow. This was even worst when I realized each http connection to Docker initiated a new SSH connection which was super slow. I am talking about 5s to just get the UI show any data. 

I felt maybe SSH is OK to be slow. So I packaged everything up in Docker, but it didn't work. Then I realized Docker's Go client doesn't have native SSH support. It uses SSH from OS. Which means that my tiny Docker image that doesn't ship with any thing doesn't work. 

So those were some of the limitation I faced. 

However, I get this question asked a lot. I really do think it's super hard to setup TLS for Docker. I don't know why it is so hard. I personally never did it and used `docker-machine` to automate the process. Then just copy the certs over. 

I think Dozzle can help to make this easier. So here is some options: 

1. **Does it have to be secure?** What kind of env do you have? You said it can't be insecure which means `tecnativa/docker-socket-proxy` is not an option. Can you help me understand if this is a corporate network that needs to be secure or something else?  With this option, Dozzle only needs to read containers and no `write` permissions. 
2.  Use `docker-machine`. With `docker-machine`, you can do something like `docker-machine create -d generic --generic-ssh-user user --generic-ip-address x.x.x.x test`. However, docker machine support has been discontinued so your experience might vary. 
3. Finally, I was thinking about this. Maybe I can create a docker image similar to `tecnativa/docker-socket-proxy`, except it generate TLS certs on the fly and places them in a directory. Then it will spin a simple proxy that forward everything from a port to `socket` file. All you would have to do is copy over the certs to Dozzle and it should work without a lot of complications. This sounds like a fun little project but it will probably take some testing so I don't know if this is something you want to do. 

Before spending a lot of time, it would be great to understand your problem. Personally, I don't think SSH will work really well. There is a [section](https://docs.docker.com/engine/security/protect-access/#ssh-tips) on this which suggest to use keep alive to speed things up. 

What are you thoughts?

**Comment by baikunz** [comment]
@amir20 Let me clarify some things !

When I said secure, I didn't meant corporate network secure. It's just for a homelab, nothing critical however I obviously don't want to expose the docker api unsecured.

Regarding SSH, maybe it was not clear enough in my first comment but my solution does not need anything more than what is already shipped in your current docker image.

I am using SSH to tunnel the socket (docker.sock) from my remote host, to my local server already hosting another docker instance. Which means I have access to two docker sockets on my local server. I can then mount both sockets in the containers that need access to the docker api (i.e. portainer).
I expected to be able to connect to another socket in dozzle but it does not seem to be possible.
I tried to bind the remote socket to a local port and use it in dozzle but it did not seem to work either.

Actually I think your idea of using `tecnativa/docker-socket-proxy` could actually work in my case. I could just mount the remote socket in the docker-socket-proxy container to expose it to dozzle. It would however be simpler to just be able to directly mount multiple sockets in the dozzle container.

**Comment by amir20** [comment]
Wow. Am I glad I didn't write any code for this haha. I definitely didn't understand what you meant. 

> I am using SSH to tunnel the socket (docker.sock) from my remote host, to my local server already hosting another docker instance. Which means I have access to two docker sockets on my local server. I can then mount both sockets in the containers that need access to the docker api (i.e. portainer).

Now that makes sense more. 

> I expected to be able to connect to another socket in dozzle but it does not seem to be possible.

Yes, not possible. Only single socket file is supported. Other connections have to be TCP. 

> I tried to bind the remote socket to a local port and use it in dozzle but it did not seem to work either.

I am not sure what you mean by this. But that should work. I assume you mean with `--remote-host`. I actually tested it. But it depends of your exact setup I guess. 

>  I could just mount the remote socket in the docker-socket-proxy container to expose it to dozzle. It would however be simpler to just be able to directly mount multiple sockets in the dozzle container.

Yes, that should work. 

> It would however be simpler to just be able to directly mount multiple sockets in the dozzle container.

I guess it depends how many other people need it. I haven't seen a lot of people wanting multiple socket files. 

I think your best bet is to somehow have a proxy that connects a TCP port to the socket file. `tecnativa/docker-socket-proxy` does that. But I am sure there are other options. 

I was kinda excited to option 3, though. :P But sounds like you don't need that.

**Comment by baikunz** [comment]
Indeed ! My bad for not wording my request properly !

By the way thanks to your replies I figured I overlooked my trial to directly forward the remote socket file to a tcp port.
I tried it and it failed at first but only because by default it only listen to requests emitted from 127.0.0.1 which is obviously not going to work from a container using a bridged network.
Anyway I tweaked my SSH tunnel a bit and it now works perfectly with a REMOTE_HOST env set !

I also tried proxying the socket file through `tecnativa/docker-socket-proxy` and it worked just fine too, with the added bonus of being able to restrict access to parts of the api.

Anyway, sorry for the inconvenience !

I indeed won't need option 3, sorry ;D But I'm pretty sure many others would find it useful ! I just needed a quick, dirty and relatively safe way to remotely access my docker API for which my SSH tunnel gets the job done !
